### PR TITLE
Improve/simplify tree-mending

### PR DESCRIPTION
**Pull request by CGNonofr**
Tree-mending is the opposite of tree-shaking :)

The current way of polyfilling monaco became really too hacky and not robust enough

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -10,6 +10,32 @@ The VSCode api is composed of:
...
+```bash
+npm install vscode@npm:@codingame/monaco-vscode-api
+npm install -D @types/vscode
+```
+
+⚠️ And add in your package.json ⚠️:
+```json
+{
+  "scripts": {
+    "postinstall": "monaco-treemending",
```
Comment (souporserious) [inline README.md:24]: Will it be possible to run this script another way and not in `postinstall`? For example, in my Next.js plugin when a user runs `next dev`, should `monaco-treemending` be safe to call each time the server starts using child_process exec?

Comment (CGNonofr) [reply README.md:24]: It just applies a patch so you can run it anytime, it will just fail if it was already run

What is the issue with postinstall ?

Comment (souporserious) [reply README.md:24]: Sounds good! I'd like the install story for end-users of the plugin to be as simple as possible. If anything, I could codemod the package.json when the plugin is first run.

File: demo/src/main.ts [hunk 2]
```diff
@@ -14,6 +14,9 @@ import 'monaco-editor/esm/vs/language/json/monaco.contribution'
 import { createConfiguredEditor, synchronizeJsonSchemas, createModelReference } from 'vscode/monaco'
 import { SimpleTextFileSystemProvider, registerFileSystemOverlay } from 'vscode/service-override/files'
 import * as vscode from 'vscode'
+import { ILogService, LogLevel, StandaloneServices } from 'vscode/services'
+
+StandaloneServices.get(ILogService).setLevel(LogLevel.Off)
```
Comment (kaisalmen) [inline demo/src/main.ts:19]: This should demonstrate what was asked in some issue the other week, right?

File: .gitignore [hunk 3]
```diff
@@ -1,4 +1,5 @@
 node_modules/
 vscode/
 vscode-default-extensions/
-dist/
\ No newline at end of file
+dist/
+monaco-editor-treemending.patch
```
[the remark this answers is not in the document]
Comment (kaisalmen) [reply .gitignore:5]: > So, you have to re-create this patch with every new monaco-editor version, right?

I understood how it works a now, see later comment.

**Review by kaisalmen — COMMENTED** [review]
Overall, this seems like a good idea. 👍 Can you release a pre-release version? I would like to give that a try and test it. Thanks you.

File: demo/src/main.ts (hunk 2 above)
Comment (CGNonofr) [reply demo/src/main.ts:19]: Precisely!

File: package.json [hunk 4]
```diff
@@ -215,11 +216,9 @@
     "rollup-plugin-dts": "^5.3.0",
     "rollup-plugin-external-assets": "^3.0.1",
     "rollup-plugin-glob-import": "^0.5.0",
-    "rollup-plugin-styles": "^4.0.0",
     "semver": "^7.3.8",
     "ts-node": "^10.9.1",
     "typescript": "5.0.3",
```
[the remark this answers is not in the document]
Comment (CGNonofr) [reply package.json:221]: Sure!

File: scripts/install-vscode [hunk 5]
```diff
@@ -32,10 +34,51 @@ yarn install --ignore-scripts
 echo "Patching vscode..."
 patch -p1 < $patch_file
 
+# build editor without treeshaking to generate the treemending patch
```
[the remark this answers is not in the document]
Comment (CGNonofr) [reply scripts/install-vscode:37]: Oh yes, that's the whole point of this change, to not have to think about it and to not have issues related to treeshaking anymore

[the remark this answers is not in the document]
Comment (kaisalmen) [reply scripts/install-vscode:37]: Yes, in the beginning I didn't know this patch generation is fully automated. I realized this when looking at the changes.

File: .gitignore (hunk 3 above)
[the remark this answers is not in the document]
Comment (CGNonofr) [reply .gitignore:5]: > Do really want to put it in the top-level dir?

I anticipate edge cases where the user will want to use the patch by hands for some reason so I want it to be easily accessible

> Was generally thinking if you want to move all the patch stuff / vscode / vscode-default-extensions into some sub directory

Why not about vscode and vscode-default-extensions, but I'm not sure what to name this directory... what do you suggest?

> Plus, add a purge script in the package.json so you can reset fully as clean no longer removes vscode dir if the version did not change. WDYT?

Sure why not! Is there a command to clean the node_modules? I don't find any

File: package.json [hunk 6]
```diff
@@ -163,19 +156,24 @@
...
     "clean": "rm -rf dist/",
-    "compile-webpack-loader": "tsc --target es2020 -lib es2020 --module commonjs --skipLibCheck --moduleResolution node --declaration --outDir dist src/webpack-loader.ts && mv dist/webpack-loader.js dist/webpack-loader.cjs",
-    "compile-rollup-plugins": "rollup --config rollup/rollup.rollup-plugins.config.ts --configPlugin 'typescript={tsconfig: `tsconfig.rollup-config.json`, include: [`./rollup/rollup.rollup-plugins.config.ts`]}'",
-    "preprepare": "./scripts/install-vscode",
+    "compile-treemending-script": "rollup --config rollup/rollup.treemending-script.config.ts --configPlugin 'typescript={tsconfig: `tsconfig.rollup-config-treemending-script.json`, include: [`./rollup/rollup.treemending-script.config.ts`]}'",
+    "compile-rollup-plugins": "rollup --config rollup/rollup.rollup-plugins.config.ts --configPlugin 'typescript={tsconfig: `tsconfig.rollup-config-plugins.json`, include: [`./rollup/rollup.rollup-plugins.config.ts`]}'",
+    "preprepare": "./scripts/install-vscode && node --loader ts-node/esm src/monaco-treemending.ts",
```
[the remark this answers is not in the document]
Comment (CGNonofr) [reply package.json:169]: How can it happen?

I'm not sure how to handle it, should we create a file to know the patch was already applied?

[the remark this answers is not in the document]
Comment (kaisalmen) [reply package.json:169]: Something like an expected hash of all files and store that. If it matches do not apply. But, if it just generates work, then leave it as is.

[the remark this answers is not in the document]
Comment (kaisalmen) [reply package.json:169]: Just leave it as is. There is no low-hanging fruit and therefore it becomes overkill, I agree.

File: .gitignore (hunk 3 above)
[the remark this answers is not in the document]
Comment (kaisalmen) [reply .gitignore:5]: > I anticipate edge cases where the user will want to use the patch by hands for some reason so I want it to be easily accessible

Ok, makes sense.

> Why not about vscode and vscode-default-extensions, but I'm not sure what to name this directory... what do you suggest?

What about moving `rollup`, `scripts`, `vscode` and `vscode-default-extensions` along with all rollup configs into a sub-dir called `build` or `transform` or something better fitting?
I don't want to ask for unnecessary work here, but it seems that gathering related stuff in specific places could make sense long-term.

> Sure why not! Is there a command to clean the node_modules? I don't find any

This repo requires Linux, so just use brute force: `rm -fr <dirs>`

[the remark this answers is not in the document]
Comment (kaisalmen) [reply .gitignore:5]: This seems to do the job (add `-n` for dry-run):
```
git clean -f -X -d
```

**Comment by kaisalmen** [comment]
From https://github.com/TypeFox/monaco-languageclient/pull/485 so it is not forgotton:

This no longer works:
```javascript
const vscodeDocument = vscode.workspace.textDocuments[0];
```

and replacement
```javascript 
const mainVscodeDocument = await vscode.workspace.openTextDocument({
```

throws an error:
image

File: package.json (hunk 4 above)
[the remark this answers is not in the document]
Comment (CGNonofr) [reply package.json:221]: Hum can't for now, it conflict with a dependency and I need to overrides it

**Comment by CGNonofr** [comment]
> From [TypeFox/monaco-languageclient#485](https://github.com/TypeFox/monaco-languageclient/pull/485) so it is not forgotton:
> 
> This no longer works:
> 
> ```js
> const vscodeDocument = vscode.workspace.textDocuments[0];
> ```
> 
> and replacement
> 
> ```js
> const vscodeDocument = await vscode.workspace.openTextDocument({
>     content,
>     language: languageId
> });
> ```
> 
> throws an error: image

if by `This no longer works` you means this is undefined, it's probably because you call it too soon

`vscode.workspace.openTextDocument` not working on untitled looks legit

**Comment by CGNonofr** [comment]
v1.78.5-next.0 published

**Comment by kaisalmen** [comment]
> if by This no longer works you means this is undefined, it's probably because you call it too soon

Then, it is likely a Layer 8 issue

**Comment by CGNonofr** [comment]
> > if by This no longer works you means this is undefined, it's probably because you call it too soon
> 
> Then, it is likely a Layer 8 issue

Ahah never heard that network joke before :+1:

**Comment by kaisalmen** [comment]
Problem above is a non-issue and indeed a user error, but when I upgrade to version `1.78.5-next.0` in monaco-languageclient both verification builds for webpack and vite break:
https://github.com/TypeFox/monaco-languageclient/actions/runs/4865173108
Any idea?

**Comment by CGNonofr** [comment]
> Problem above is a non-issue and indeed a user error, but when I upgrade to version `1.78.5-next.0` in monaco-languageclient both verification builds for webpack and vite break: https://github.com/TypeFox/monaco-languageclient/actions/runs/4865173108 Any idea?

It looks very like the patch wasn't applied

**Comment by kaisalmen** [comment]
> It looks very like the patch wasn't applied

Note to myself: RTFM

File: package.json (hunk 4 above)
[the remark this answers is not in the document]
Comment (kaisalmen) [reply package.json:221]: Ok

File: .gitignore (hunk 3 above)
[the remark this answers is not in the document]
Comment (kaisalmen) [reply .gitignore:5]: I will create a small PR after this is merged and add this:
```json
    "reset:repo": "git clean -f -X -d"
```

[the remark this answers is not in the document]
Comment (CGNonofr) [reply .gitignore:5]: Do not hesitate to add it here

[the remark this answers is not in the document]
Comment (kaisalmen) [reply .gitignore:5]: Pushed now on the correct branch. I reverted the change on `view-services`

**Review by kaisalmen — APPROVED** [review]
From my point of view this is ready to go. Everything works with monaco-languageclient. I checked all examples again there just now.

**Comment by kaisalmen** [comment]
> Monaco 0.38 was released yesterday 😁

Will this go into vscode 1.78.0 or onto 1.79.0?

**Comment by CGNonofr** [comment]
> > Monaco 0.38 was released yesterday grin
> 
> Will this go into vscode 1.78.0 or onto 1.79.0?

the editor references https://github.com/microsoft/vscode/commit/5474fca23cc6723a6f72975fbf00672c331fed2c so it will be a 1.79

**Pull request merged**